rescript/tests/test_get_unite.py

```diff
@@ -0,0 +1,85 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2019-2023, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import pkg_resources
+import tempfile
+from qiime2.plugin.testing import TestPluginBase
+from rescript.get_unite import (
+    UNITE_DOIS,
+    _unite_get_url,
+    _unite_get_tgz,
+    _unite_get_artifacts,
+    get_unite_data,
+)
+
+from urllib.request import urlopen
+from urllib.error import HTTPError
+from unittest.mock import patch
+
+
+class TestGetUNITE(TestPluginBase):
+    package = "rescript.tests"
+
+    def setUp(self):
+        super().setUp()
+        self.unitefile = pkg_resources.resource_filename(
+            "rescript.tests", "data/unitefile.tgz"
+        )
+
+    # Requires internet access
+    def test_unite_get_url(self):
+        # for all combinations...
+        for v in UNITE_DOIS.keys():
+            for tg in UNITE_DOIS[v].keys():
+                for s in UNITE_DOIS[v][tg].keys():
+                    # ... try to get the URL
+                    try:
+                        url = _unite_get_url(v, tg, s)
+                        urlopen(url)
+                    except HTTPError:
+                        raise ValueError("No URL for combo: " + v + tg + s)
+
+    # Requires internet access
+    def test_unite_get_tgz(self):
+        # Download a single, small, unrelated file for testing
+        url = "https://files.plutof.ut.ee/doi/C9/F6/C9F687C997F72F674AA539CB80BF5D5BF6D1F402A2ACF840B20322850D3DFBA4.zip"  # noqa E501
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            # Test working URL
+            _unite_get_tgz(url, tmpdirname)
+            # Test bad url
+            with self.assertRaises(ValueError):
+                _unite_get_tgz("https://files.plutof.ut.ee/nope", tmpdirname)
```

Comment:
This test fails, because above I catch the ValueError and print "File incomplete, on try"

After 10 failed downloads, I need a way to fail harder

Reply:
As you already test for the DOIs in `test_unite_get_url`, not sure there is a need to do a related test again. Unless I am missing something. That being said, you can also try as per [this SILVA example](https://github.com/bokulich-lab/RESCRIPt/blob/946728467532f066c8bb750c3d8d1b8af02773b9/rescript/tests/test_get_data.py#L42C7-L42C7). Basically, just add `self.assertTrue(True)` after the command runs.

You an also use `patch` for ` _unite_get_tgz` here too. Just feed a tiny, similarly formatted tgz file.

Reply:
Oh yeah, the URL passed to this function should never be bad.

Should I remove this 'bad url' test like I had originally?

Reply:
Yeah I'd remove the bad url.

Reply:
Ok, that almost works. I was testing the wrong thing :-/

Sometimes, their API ends early, and I have a `File download incomplete` loop that catches it.

After catching the 'failed to download' error 10 times, I forgot to raise a final error, which I can now test for with this bad URL.

Let me push this now!